src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1130,24 +1145,24 @@ ucs_status_t uct_ud_ep_flush(uct_ep_h ep_h, unsigned flags,
 
     uct_ud_enter(iface);
 
-    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
-        uct_ud_ep_purge(ep, UCS_ERR_CANCELED);
-        /* FIXME make flush(CANCEL) operation truly non-blocking and wait until
-         * all of the outstanding sends are completed. Without this, zero-copy
-         * sends which are still on the QP could be reported as completed which
-         * can lead to sending corrupt data, or local access error. */
-        status = UCS_OK;
-        goto out;
+    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL) &&
+        !(ep->flags & UCT_UD_EP_FLAG_FLUSH_CANCEL)) {
+        uct_ud_iface_dispatch_async_comps(iface, ep);
+        uct_ud_ep_purge(ep, UCS_ERR_CANCELED, 0);
+        ep->flags |= UCT_UD_EP_FLAG_FLUSH_CANCEL;
     }
 
     if (ucs_unlikely(uct_ud_iface_has_pending_async_ev(iface))) {
         status = UCS_ERR_NO_RESOURCE;
         goto out;
     }
 
-    status = uct_ud_ep_flush_nolock(iface, ep, comp);
+    status = uct_ud_ep_flush_nolock(iface, ep, flags, comp);
     if (status == UCS_OK) {
         UCT_TL_EP_STAT_FLUSH(&ep->super);
+        if (flags & UCT_UD_EP_FLAG_FLUSH_CANCEL) {
```

Comment:
can we remove that if (i.e expect window is empty if uct_ud_ep_flush_nolock returned OK, regardless of flags)